ssl/ssl_sess.c

```diff
@@ -59,17 +59,37 @@ __owur static int timeoutcmp(SSL_SESSION *a, SSL_SESSION *b)
  */
 void ssl_session_calculate_timeout(SSL_SESSION *ss)
 {
-#ifndef __DJGPP__ /* time_t is unsigned on djgpp */
-    /* Force positive timeout */
-    if (ss->timeout < 0)
-        ss->timeout = 0;
+    time_t tmax = (time_t)-1;
+
+    if (sizeof(time_t) == 8) {
+        uint64_t overflow;
+
+#ifndef __DJGPP__ /* time_t is unsigned on djgpp, it's signed anywhere else */
+        tmax = (time_t)((uint64_t)tmax >> 1);
 #endif
-    ss->calc_timeout = ss->time + ss->timeout;
-    /*
-     * |timeout| is always zero or positive, so the check for
-     * overflow only needs to consider if |time| is positive
-     */
-    ss->timeout_ovf = ss->time > 0 && ss->calc_timeout < ss->time;
+        overflow = (uint64_t)tmax - (uint64_t)ss->time;
+        if (ss->timeout > (time_t)overflow) {
+            ss->timeout_ovf |= 1;
+            ss->calc_timeout = ss->timeout - (time_t)overflow;
+        } else {
+            ss->calc_timeout = ss->time + ss->timeout;
+        }
+    } else {
+        uint32_t overflow;
+
+#ifndef __DJGPP__ /* time_t is unsigned on djgp, it's signed anywhere else */
+        tmax = (time_t)((uint32_t)tmax >> 1);
+#endif
+        overflow = (uint32_t)tmax - (uint32_t)ss->time;
```

Comment:
The logic for calculating overflow and setting calc_timeout is duplicated for `32-bit` and `64-bit` time_t. Consider refactoring this into a helper function to reduce redundancy.

Reply:
To help you, here is the code for it:

```
#define CALCULATE_TIMEOUT(ss, tmax, type) \
    do { \
        type overflow; \
        tmax = (time_t)((type)tmax >> 1); \
        overflow = (type)tmax - (type)ss->time; \
        if (ss->timeout > (time_t)overflow) { \
            ss->timeout_ovf |= 1; \
            ss->calc_timeout = ss->timeout - (time_t)overflow; \
        } else { \
            ss->calc_timeout = ss->time + ss->timeout; \
        } \
    } while (0)

/*
 * Calculates effective timeout, saving overflow state
 * Locking must be done by the caller of this function
 */
void ssl_session_calculate_timeout(SSL_SESSION *ss)
{
    time_t tmax = (time_t)-1;

    if (sizeof(time_t) == 8) {
        CALCULATE_TIMEOUT(ss, tmax, uint64_t);
    } else {
        CALCULATE_TIMEOUT(ss, tmax, uint32_t);
    }

    /*
     * N.B. Realistic overflow can only occur in our lifetimes on a
     *      32-bit machine with signed time_t, in January 2038.
     *      However, There are no controls to limit the |timeout|
     *      value, except to keep it positive.
     */
}

```

Reply:
The DJGPP ifdefed part would have to be put outside of the macro but this still looks like a good suggestion.

Reply:
I like it too.